### ISSUE TITLE
Order of exported classes

### ISSUE THREAD

**Issue by Crozin** [reinforced/Reinforced.Typings#21]
Hi.
There is an issue with order of generated `export class`es. Reinforced.Typings builds following code:

```
module MyModule {
  export class Abc extends MyModule.Def { }
  export class Def { }
}
```

And this causes TypeScript error:

```
error TS2690: A class must be declared after its base class.
```

**Comment by pavel-b-novikov** [comment]
Wow, we've got critical.
Unfortunately, I can take a look at it only next week. Right now I'm moving from country to country

**Comment by pavel-b-novikov** [comment]
As quick workaround I can suggest you to instruct RT to move your base class to another file

**Comment by attilabicsko** [comment]
Hi,

Is there a way to configure RT to generate only some (base) classes to another file, but leave all the other classes in a different, single file? Can you please give an example on how to workaround this? Or will this issue be fixed at some point? Thanks in advance!

**Comment by pavel-b-novikov** [comment]
@attilabicsko , here is your sample to accomplish your task with fluent configuration

https://gist.github.com/pavel-b-novikov/fc784a5ebf34a5c5a56a8318a086aafd

**Comment by attilabicsko** [comment]
@pavel-b-novikov Thank you for your quick reply. I tried that, but this solution seem to break the TsClass attribute's Namespace property, it just got ignored and I rely on that setting. Also, I'm generating a lot of classes, so maintaining a list of base types and other types in fluent configuration is unfortunately not an option (tried with reflection though, but that created different problems). For now, I ended up using the TsFile attribute explicitly for each class attributed with TsClass, this seem to solve the issue for me. Thanks again!

**Comment by pavel-b-novikov** [comment]
@attilabicsko I'm glad that you've solved your issue and sorry for not pointing to TsFile - really, I forgot because I'm using fluent configuration in my projects. 

And yes, I want to apologize to eveyone for too long new version development - I have much more important tasks now.

**Comment by konrad-powazka** [comment]
I had the same issue. As a workaround a created a little helper method which for given types gets their inheritance hierarchy ordered so that TypeScript compiler will not moan. I haven't tested it thoroughly, but for my project it works so far.

```csharp
public static void Configure(ConfigurationBuilder builder)
{
    var typesToExport = GetTypesToExport(); // Get types you need to export to TS
    var orderedTypesToExport = GetOrderedInheritanceHierarchy(typesToExport);
    builder.ExportAsClasses(orderedTypesToExport);
}

private static IEnumerable<Type> GetOrderedInheritanceHierarchy(IEnumerable<Type> types)
{
    var unorderedTypesHierarchy =
        types.SelectMany(GetTypeInheritanceChain).Where(t => t != typeof(object)).Distinct().ToList();

    var typesOrderedInLastIteration = unorderedTypesHierarchy.Where(t => t.BaseType == typeof(object)).ToList();
    var orderedTypesHierarchy = new List<Type>();
    var typesLeftToOrder = new HashSet<Type>(unorderedTypesHierarchy);

    while (typesLeftToOrder.Any())
    {
        foreach (var typeOrderedInLastIteration in typesOrderedInLastIteration)
        {
            typesLeftToOrder.Remove(typeOrderedInLastIteration);
            orderedTypesHierarchy.Add(typeOrderedInLastIteration);
        }

        typesOrderedInLastIteration =
            typesLeftToOrder.Where(
                derivedType => typesOrderedInLastIteration.Any(baseType => derivedType.BaseType == baseType))
                .ToList();
    }

    Debug.Assert(!typesLeftToOrder.Any());

    return orderedTypesHierarchy;
}
```

**Comment by pavel-b-novikov** [comment]
@Crozin , Explicit export types ordering implemented according to your request in 1.30 branch. 
Please check out and run your test.
Sorry, I do not have enough time to implement honest topological sort there

**Comment by pavel-b-novikov** [comment]
Fixed in 1.3.0
